### PR TITLE
Adds customizable sound emitters for admin use

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
admin: Admins can now create sound emitters (/obj/effect/sound_emitter) that can be customized to play sounds to different people, at different volumes, and at different locations such as by z-level. They're much more versatile for events than the Play Sound commands!
/:cl:

[Short video demo can be found here.](https://youtu.be/pMMALthQEHY) Some stuff isn't current!

When I ran the Colosseum of Fools event a while ago, I used sounds I had locally to spice up the event and make it unique, but I ran into the issue of there being no easy way to play sounds in flexible ways while preventing people who don't want to hear them from being included.

Adds sound emitter objects, editable by admins through a handy UI, that allows them to play sounds in a variety of flexible ways. The following things can be manipulated by any admin by clicking on the sound emitter:
* **Label** (maptext)
* **Sound File** is chosen by navigating to the file like normal sound playing
* **Sound Volume** is a percentage between 1 and 100
* **Mode** dictates how the sound is played and can either be local (played at that location like a normal in-game sound) or direct (played directly to players like admin sounds)
* **Range** dictates how far players need to be to hear the sound; this includes Radius (all players within a specified amount of tiles), Z-Level (all players on the same z-level), and Global (all players.)

Sound emitters can be examined by any ghosts (they're invisible to normal players) to see the info about them, but only admins can edit their properties and play sounds with them.

This lets admins make very flexible sound emitters that can fulfill a variety of different purposes, like music that only people within 7 tiles can hear or a distant screech of metal that an entire search party on the derelict hears.

In addition to the Play Sound button, alt-clicking the emitter activates it.

### CODE REVIEW

**Comment by Arianya** [comment]
Does this respect player preference wrt Admin MIDIs?

**Comment by kevinz000** [comment]
pretty good

**Comment by Xhuis** [comment]
@Arianya Not at the moment since it's entirely customizable and I intend for this to be used for sounds and not music. I could make it utilize midi prefs in the case of global sounds, however, if an admin would like.

**Comment by Arianya** [comment]
I think its going to be preferable to just have them respect prefs by default, since otherwise it'll lead to complaints and so on.

People who disabled admin MIDIs did so for a reason after all.

**Comment by lzimann** [comment]
nothing more cancer than admins forcing their stupid sounds into everyone, either use instruments or admin midis preference for this

File: code/modules/admin/sound_emitter.dm [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+/obj/effect/sound_emitter/examine(mob/user)
+	..()
+	if(!isobserver(user))
+		return
+	to_chat(user, "<span class='boldnotice'>Sound File:</span> [sound_file ? sound_file : "None chosen"]")
+	to_chat(user, "<span class='boldnotice'>Mode:</span> [motus_operandi]</span>")
+	to_chat(user, "<span class='boldnotice'>Range:</span> [emitter_range]</span>")
+	to_chat(user, "<b>Sound is playing at [sound_volume]% volume.</b>")
+	if(user.client.holder)
```
Comment (optimumtact) [inline code/modules/admin/sound_emitter.dm]: please add an explicit rights check, not every downstream wants all users with an admin holder to be able to access thigns like this.

Also it's lazy coding.

**Review by optimumtact — CHANGES_REQUESTED** [review]

File: code/modules/admin/sound_emitter.dm (hunk 1 above)
Comment (Xhuis) [reply code/modules/admin/sound_emitter.dm]: I haven't messed with admin code enough to know that by heart, bear with me. I'll quickly do a check for sound rights.

**Review by optimumtact — APPROVED** [review]

File: code/modules/admin/sound_emitter.dm [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	edit_emitter(user)
+
+/obj/effect/sound_emitter/AltClick(mob/user)
+	if(check_rights_for(user.client, R_SOUNDS))
+		activate(user)
+		to_chat(user, "<span class='notice'>Sound emitter activated.</span>")
+
+/obj/effect/sound_emitter/proc/edit_emitter(mob/user)
+	var/dat = ""
+	dat += "<b>Label:</b> <a href='?src=\ref[src];edit_label=1'>[maptext ? maptext : "No label set!"]</a><br>"
```
Comment (AnturK) [inline code/modules/admin/sound_emitter.dm]: /ref's ree

Comment (optimumtact) [reply code/modules/admin/sound_emitter.dm]: what are you supposed to use instead in hrefs antur

Comment (Xhuis) [reply code/modules/admin/sound_emitter.dm]: >not coding your own framework for replacing hrefs

Comment (AnturK) [reply code/modules/admin/sound_emitter.dm]: REF() function, since 512 changes functionality.

**Pull request merged**